source/eventcore/drivers/posix/sockets.d

```diff
@@ -113,14 +113,32 @@ final class PosixEventDriverSockets(Loop : PosixEventLoop) : EventDriverSockets
 				m_loop.clearFD(sock);
 				m_loop.unregisterFD(sock, EventMask.read|EventMask.write|EventMask.status);
 				invalidateSocket();
-				on_connect(sock, ConnectStatus.unknownError);
-				return sock;
+				on_connect(StreamSocketFD.invalid, ConnectStatus.unknownError);
+				return StreamSocketFD.invalid;
 			}
 		}
 
 		return sock;
 	}
 
+	final override void cancelConnectStream(StreamSocketFD sock)
+	{
+		assert(sock != StreamSocketFD.invalid, "Invalid socket descriptor");
+		with (m_loop.m_fds[sock].streamSocket)
+		{
+			assert(state == ConnectionState.connecting,
+				"Unable to cancel connect on the socket that is not in connecting state");
+			state = ConnectionState.closed;
+			auto cb = connectCallback;
+			connectCallback = null;
+			m_loop.clearFD(sock);
+			m_loop.unregisterFD(sock, EventMask.read|EventMask.write|EventMask.status);
+			closeSocket(cast(sock_t)sock.value);
+			if (cb)
+				cb(StreamSocketFD.invalid, ConnectStatus.cancelled);
```

Comment:
The semantic for `cancelX` is to guarantee that the callback will never be called after the cancel request has been issued, so this call here should be removed. The reason is that the fiber switching logic in vibe-core can be generalized easily that way.

Reply:
ok, so I can rollback "cancelled" enum meber as well, got it.